qtacrylicmainwindow.cpp

```diff
@@ -0,0 +1,199 @@
+        update();
+        Q_EMIT tintOpacityChanged();
+    }
+}
+
+qreal QtAcrylicMainWindow::noiseOpacity() const
+{
+    return m_acrylicHelper.getNoiseOpacity();
+}
+
+void QtAcrylicMainWindow::setNoiseOpacity(const qreal value)
+{
+    if (m_acrylicHelper.getNoiseOpacity() != value) {
+        m_acrylicHelper.setNoiseOpacity(value);
+        m_acrylicHelper.updateAcrylicBrush(tintColor());
+        update();
+        Q_EMIT noiseOpacityChanged();
+    }
+}
+
+bool QtAcrylicMainWindow::frameVisible() const
+{
+    return m_frameVisible;
+}
+
+void QtAcrylicMainWindow::setFrameVisible(const bool value)
+{
+    if (m_frameVisible != value) {
+        m_frameVisible = value;
+        update();
+        Q_EMIT frameVisibleChanged();
+    }
+}
+
+QColor QtAcrylicMainWindow::frameColor() const
+{
+    return m_acrylicHelper.getFrameColor();
+}
+
+void QtAcrylicMainWindow::setFrameColor(const QColor &value)
+{
+    if (m_acrylicHelper.getFrameColor() != value) {
+        m_acrylicHelper.setFrameColor(value);
+        update();
+        Q_EMIT frameColorChanged();
+    }
+}
+
+qreal QtAcrylicMainWindow::frameThickness() const
+{
+    return m_acrylicHelper.getFrameThickness();
+}
+
+void QtAcrylicMainWindow::setFrameThickness(const qreal value)
+{
+    if (m_acrylicHelper.getFrameThickness() != value) {
+        m_acrylicHelper.setFrameThickness(value);
+        update();
+        Q_EMIT frameThicknessChanged();
+    }
+}
+
+bool QtAcrylicMainWindow::acrylicEnabled() const
+{
+    return m_acrylicEnabled;
+}
+
+void QtAcrylicMainWindow::setAcrylicEnabled(const bool value)
+{
+    if (m_acrylicEnabled != value) {
+        m_acrylicEnabled = value;
+        if (m_acrylicEnabled) {
+            connect(&m_acrylicHelper, &QtAcrylicEffectHelper::needsRepaint, this, qOverload<>(&QtAcrylicMainWindow::update));
+        } else {        
+            disconnect(&m_acrylicHelper, &QtAcrylicEffectHelper::needsRepaint, this, qOverload<>(&QtAcrylicMainWindow::update));
+        }
+        if (m_inited) {
+            Utilities::setBlurEffectEnabled(windowHandle(), m_acrylicEnabled);
+        }
+        setAutoFillBackground(!m_acrylicEnabled);
+        setAttribute(Qt::WA_NoSystemBackground, m_acrylicEnabled);
+        setAttribute(Qt::WA_OpaquePaintEvent, m_acrylicEnabled);
+        setBackgroundRole(m_acrylicEnabled ? QPalette::Base : QPalette::Window);
+        update();
+        Q_EMIT acrylicEnabledChanged();
+    }
+}
+
+void QtAcrylicMainWindow::showEvent(QShowEvent *event)
+{
+    QMainWindow::showEvent(event);
+    updateContentMargin();
+    if (!m_inited) {
+        FramelessWindowsManager::addWindow(windowHandle());
+        m_acrylicHelper.install(windowHandle());
+        m_inited = true;
+    }
+    m_acrylicHelper.updateAcrylicBrush(tintColor());
+    setAcrylicEnabled(acrylicEnabled());
+}
+
+void QtAcrylicMainWindow::updateContentMargin()
+{
+    const qreal m = isMaximized() ? 0.0 : 1.0 / windowHandle()->devicePixelRatio();
```

Comment:
you can use `devicePixelRatioF()` to replace `windowHandle()->devicePixelRatio()`, because windowHandle() maybe null